src/search.jl

```diff
@@ -544,3 +546,410 @@ function execute(search::GeneticAlgorithmSearch, problem)
         end
     end
 end
+
+
+#=
+Julia - GSOC
+
+Ahmed Madbouly
+
+Implementation for AND-OR, Online DFS, LRTA
+=#
+
+abstract type AbstractProblemStruct end;
+
+mutable struct ProblemStruct <: AbstractProblemStruct
+    initial::String
+    goal::Nullable{String}
+
+    function ProblemStruct(initial_state::String; goal_state::Union{Void, String}=nothing)
+        return new(initial_state, Nullable{String}(goal_state));
+    end
+
+end
+
+#=
+    OnlineDFSAgentProgram is a online depth first search agent (Fig. 4.21)
+=#
+mutable struct OnlineDFSAgentProgram <: AgentProgram
+    result::Dict
+    untried::Dict
+    unbacktracked::Dict
+    state::Nullable{String}
+    action::Nullable{String}
+    problem::AbstractProblemStruct
+
+    function OnlineDFSAgentProgram{T <: AbstractProblemStruct}(problem::T)
+        return new(Dict(), Dict(), Dict(), Nullable{String}(), Nullable{String}(), problem);
+    end
+end
+
+struct MemoizedFunction
+    f::Function     #original function
+    values::Dict{Tuple{Vararg}, Any}
+
+    function MemoizedFunction(f::Function)
+        return new(f, Dict{Tuple{Vararg}, Any}());
+    end
+end
+
+struct Graph{N}
+    dict::Dict{N, Any}
+    locations::Dict{N, Tuple{Any, Any}}
+    directed::Bool
+
+    function Graph{N}(;dict::Union{Void, Dict{N, }}=nothing, locations::Union{Void, Dict{N, Tuple{Any, Any}}}=nothing, directed::Bool=true) where N
+        local ng::Graph;
+        if ((typeof(dict) <: Void) && (typeof(locations) <: Void))
+            ng = new(Dict{Any, Any}(), Dict{Any, Tuple{Any, Any}}(), Bool(directed));
+        elseif (typeof(locations) <: Void)
+            ng = new(Dict{eltype(dict.keys), Any}(dict), Dict{Any, Tuple{Any, Any}}(), Bool(directed));
+        else
+            ng = new(Dict{eltype(dict.keys), Any}(dict), Dict{eltype(locations.keys), Tuple{Any, Any}}(locations), Bool(directed));
+        end
+        if (!ng.directed)
+            make_undirected(ng);
+        end
+        return ng;
+    end
+
+    function Graph{N}(graph::Graph{N}) where N
+        return new(Dict{Any, Any}(graph.dict), Dict{String, Tuple{Any, Any}}(graph.locations), Bool(graph.directed));
+    end
+end
+
+struct OnlineSearchProblem <: AbstractProblemStruct
+    initial::String
+    goal::String
+    graph::Graph
+    least_costs::Dict
+    h::Function
+
```

Comment:
Very generic types. Please align the code to the general types defined in this file.

Reply:
1. I think a generic guideline will be to complete only LRTA now as you have added full test cases for them. Remove the others and submit as a separate PR later. 
2. Change the code to reflect the project's general type system and data structure usage. Before using any new Data Structures make sure the older ones can be enhanced properly. 
3. The changes suggested are large and affect the complete submission. So would suggest reworking on it. The inline comments are only examples not exhaustive.